### DISCUSSION TITLE
Accessing errors outside input field scope in a performant manner

### DISCUSSION THREAD

**Discussion by Thomas-Smyth** [react-hook-form/react-hook-form#8330] [Q&A] [answered]
## Background
I have been investigating switching from Formik to React Hook Forms to attempt to resolve performance issues I am experiencing with Formik as a result of its tendency to re-render the whole form whenever its state changes, due to each field in the form accessing it via a shared context. Previously, this has not been a problem as our existing forms are small enough that the re-renders are manageable, however, the form I am currently working on is much larger and more complex and as a result, the re-renders are not manageable - even when `React.memo` is used to prevent the computationally heavy fields from re-rendering.

On some of our forms, including the one I am currently working on, we rely on accessing errors outside of the scope of the associated field. Our use cases for this include disabling buttons when errors are present in fields or flagging collapsible sections in our form if there are errors present in fields within them.

In Formik, this is really easy to achieve by accessing Formik's context anywhere within the form and checking for errors within any of the relevant fields, as seen below.
```
const { errors } = useFormikContext();

return <button disabled={getIn(errors, 'fieldName')}>Do something</button>;
```

In React Hook Forms, as far as my experimenting has revealed, this is not easily achieved and can result in a complete loss in the performance improvements that were gained by switching to React Hook Form.

## The Problem
https://codesandbox.io/s/purple-violet-zmcz3n?file=/src/App.js

In the above CodeSandbox, there is a basic example of a form that replicates one of my use cases.

The example contains two interchangeable components, the latter of which is enabled by default:
 * `<SubmitButtonWithoutErrorBox />` - A submit button.
 * `<SubmitButtonWithErrorBox />`- A submit button that displays the error from the first field, caused if no value is entered into the field when the form is submitted, above it in a red box.

When the first of these components is used, all the fields in the form do not re-render when another field is blurred, however, when the second of these components is used, all the fields in the form re-render when another field is blurred. Based on the difference between these two components, the cause of this seems to be a consequence of accessing the `errors` object in the `formState` object. As this is a requirement in my use case, almost all performance improvements we expected to gain by switching to React Hook Form are lost.

## Possible Cause
Without diving too deeply into the source of React Hook Form, I hypothesise that accessing the `errors` object causes it or the `formState` object to be lazy-loaded in the form context for the component, but as an unintended consequence of this all components that access the form context, even if indirectly via the `useController` hook, also re-render even if they do not access the `formState` object. In the example above, this is all fields as they all access the form context to get the `control` object, which would be quite onerous to pass around as a prop in the form I'm developing.

## Attempted Workaround
As a possible workaround to this issue, I have tried to use `React.memo` [as suggested by the docs](https://react-hook-form.com/advanced-usage#FormProviderPerformance) to limit re-renders to only when relevant values in the context change, however, as the form I'm developing already does something similar for Formik, with a limited performance improvement, not much is gained. Furthermore, it will probably be quite onerous to write the dependencies into `React.memo` for all the components that need to access the form state.

## Questions
 * Is there any other workarounds you can suggest that I may have missed?
 * Is there a possibility that a hooksimilar to `useWatch` could be added to access not only a field's value, but also its `error` and `touched` values outside the scope of the field in a performant manner?

**Comment by bluebill1049** [comment, marked as answer]
have you considered using https://react-hook-form.com/api/useformstate instead?

**Reply by Thomas-Smyth** [reply]
Ah no, I somehow missed that.

Switching that looks to have solved the issue I was experiencing. Is there any reason why the behaviour of the two different contexts differs so much?

**Reply by bluebill1049** [reply]
all good.

> This custom hook allows you to subscribe to each form state, and isolate the re-render at the custom hook level. It has its scope in terms of form state subscription, so it would not affect other useFormState and useForm. Using this hook can reduce the re-render impact on large and complex form application.

**Reply by Thomas-Smyth** [reply]
Thanks.
